Add task cancellation mechanism

imotov opened the pull request:
Only tasks that extend CancellableTask can be cancelled using this mechanism. If a cancellable task has children it can elect to cancel all child tasks as well. In this case a special ban parent request is sent to all nodes. This request does two things: 1) it prevents any tasks with the banned parent task from being started, and 2) it cancels all currently running tasks that have the banned task as a parent. The ban is lifted as soon as the coordinating node notifies all other nodes that the cancelled task has finished executing. If the coordinating node leaves the cluster before it has a chance to lift its bans, all bans set by this coordinating node are automatically removed.

As an option a task can elect to automatically cancel all child tasks if their parent task was running on a node that just left the cluster. This option makes sense for cancellable heavy tasks that have no side-effects and only return results to the coordinating node. With the coordinating node gone, it doesn't make sense to run such tasks any longer since their results will be most likely discarded.

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+        if (lastDiscoveryNodes == null || event.nodesRemoved())  {
+            synchronized (banedParents) {
+                lastDiscoveryNodes = event.state().getNodes();
+                if (banedParents.isEmpty() == false) {
```
we can skip this check right? It will just mean we pull and empty iterator?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+                    // Remove all bans that were registered by nodes that are no longer in the cluster state
+                    Iterator<Tuple<String, Long>> banIterator = banedParents.keySet().iterator();
+                    while (banIterator.hasNext()) {
+                        Tuple<String, Long> nodeAndTaskId = banIterator.next();
```
can we please unpack the tuple right away instead of using v1 v2? just easier to read

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+            for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+                Task task = taskEntry.getValue();
+                String parent = task.getParentNode();
+                if (parent != null && lastDiscoveryNodes.nodeExists(parent) == false && task instanceof CancellableTask) {
```
just a nitpick but I think the instanceOf check should go second?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+        for (Map.Entry<Long, Task> taskEntry : tasks.entrySet()) {
+            Task task = taskEntry.getValue();
+            if (parentId == task.getParentId() && parentNode.equals(task.getParentNode())) {
+                assert task instanceof CancellableTask;
```
I think that assert is pretty unnecessary since you get a much better exception if you let it run into the classcast below? I think you should make the cast outside of the catch block instead and let if it fail hard all the time?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -19,34 +19,44 @@
         super(settings);
     }
 
+    public DiscoveryNodes lastDiscoveryNodes = null;
```
can we make this private? I also wonder if we should initialize with `DiscoveryNodes#EMPTY_NODES` instead of null that would safe us a null check here and there?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -54,8 +64,28 @@ public Task register(String type, String action, TransportRequest request) {
             }
+            try {
+                task.onRegistered();
+            } catch (Throwable t) {
```
can we make this fail hard?I don't think the logging buys us anything?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -54,8 +64,28 @@ public Task register(String type, String action, TransportRequest request) {
+                String reason = banedParents.get(new Tuple<>(task.getParentNode(), task.getParentId()));
+                if (reason != null) {
+                    try {
+                        assert task instanceof CancellableTask;
```
also here the class cast excep we get the line below is much more informative that the assert so we can skip the assert?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -65,7 +95,15 @@ public Task register(String type, String action, TransportRequest request) {
+        Task removedTask = tasks.remove(task.getId());
+        if (removedTask != null) {
+            try {
+                removedTask.onUnregistered();
```
I also think we should not catch the excep here.

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+     *
+     * Will be used in task manager stats and for debugging.
+     */
+    public int banCount() {
```
`s/banCount/getBanCount"?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -74,4 +112,86 @@ public Task unregister(Task task) {
+                    if (cancellableTask.cancelOnParentLeaving()) {
+                        try {
+                            cancellableTask.cancel("Coordinating node [" + parent + "] left the cluster");
+                        } catch (Throwable ex) {
```
either we let it blow up or we suppress the exception but I think logging is wrong here there should not be an exception?

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -0,0 +1,201 @@
+                setBanOnNodes(reason, task);
+                return task.taskInfo(clusterService.localNode(), false);
+            } else {
+                throw new TaskOperationException(task, "already cancelled");
```
make it an IllegalStateException please

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -0,0 +1,201 @@
+            if (cancellableTask.cancel(request.reason())) {
+                return cancellableTask.taskInfo(clusterService.localNode(), false);
+            } else {
+                throw new TaskOperationException(task, "already cancelled");
```
make it an IllegalStateException please

s1monw commented on core/src/main/java/org/elasticsearch/ElasticsearchException.java:
```diff
@@ -613,7 +613,9 @@ public String toString() {
-        QUERY_SHARD_EXCEPTION(org.elasticsearch.index.query.QueryShardException.class, org.elasticsearch.index.query.QueryShardException::new, 141);
+        QUERY_SHARD_EXCEPTION(org.elasticsearch.index.query.QueryShardException.class, org.elasticsearch.index.query.QueryShardException::new, 141),
+        TASK_CANCELLED_EXCEPTION(org.elasticsearch.tasks.TaskCancelledException.class, org.elasticsearch.tasks.TaskCancelledException::new, 142),
+        TASK_OPERATION_EXCEPTION(org.elasticsearch.tasks.TaskOperationException.class, org.elasticsearch.tasks.TaskOperationException::new, 143);
```
please remove TaskOperatoinException and replace it with IllegalStateException which is the right Except here since it's only used as a such

s1monw replied:
I also left comments regarding the `TaskCancelledException` which I think we also should not add

s1monw commented on core/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -54,8 +64,28 @@ public Task register(String type, String action, TransportRequest request) {
+                    try {
+                        assert task instanceof CancellableTask;
+                        ((CancellableTask) task).cancel(reason);
+                        throw new TaskCancelledException("Task cancelled before it started: {}", reason);
```
this is also an IllegalStateException?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+        return reasonAndListener.v1();
+    }
+
+    public void check() {
```
can we rename this to `boolean isCanceled()` and then instead of the exception just return a boolean? I think it would be more intuitive and we really don't need yet another exception

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public boolean match(Task task) {
+        if (super.match(task) == false) {
```
`return super.match(task) && task instanceof CancellableTask;`?

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * A request to cancel tasks
+ */
+public class CancelTasksRequest extends BaseTasksRequest<CancelTasksRequest> {
```
hmm maybe another change but can't we use writeable as the interface rather than streamable?

s1monw replied:
oh nevermind I know why it's Streamable since it's send as a request...

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -0,0 +1,77 @@
+        return task instanceof CancellableTask;
+    }
+
+    public CancelTasksRequest reason(String reason) {
```
`setReason`/`getReason`?

imotov replied:
Did we switch to getFoo/setFoo pattern in ActionRequests? I cannot find any examples.

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+
+    private static final Tuple<String, CancellationListener> TASK_FINISHED_MARKER = new Tuple<>("task finished", null);
+
+    private static ThreadLocal<CancellableTask> current = new ThreadLocal<>();
```
is this a leftover? I don't see where this is used outside of tests? and even there I think it's a huge overkill. Can we please remove this entirely. If you really need stuff like this for testing then look at `ThreadContext#setTransient` which you get from a threadpool

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+
+    private AtomicReference<Tuple<String, CancellationListener>> cancellationReasonAndListener = new AtomicReference<>();
+
+    public CancellableTask(long id, String type, String action, Provider<String> description, boolean hasChildren) {
```
I think this is a mistake here you didn't want to pull in some guice class here right? `+import org.elasticsearch.common.inject.Provider;`

Can't we just make the privder be a string instead? and if necessary it should be a `Supplier`

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+            try {
+                doCancel(reason);
+            } catch (Throwable ex) {
+                // What to do?
```
what about barf?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+        current.get().check();
+    }
+
+    private AtomicReference<Tuple<String, CancellationListener>> cancellationReasonAndListener = new AtomicReference<>();
```
this must be final?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    public interface CancellationListener {
```
I think we can just use `Runnable` instead of a new interface?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,128 @@
+        return hasChildren;
+    }
+
+    public final boolean cancel(String reason, CancellationListener listener) {
```
I don't think we should keep any state in here at all and instead move everything we keep on state in the TaskManager. That way we don't need the listener on this class and also maybe we can skip it entirely if we keep state on the TaskManager what has children and what is cancelable?

s1monw commented:
I left a bunch of comments @imotov

imotov commented:
@s1monw I still need to come up with a better task selection/matching approach in TransportTasksAction, but I would really appreciate it if you could take a look at the refactoring that I have done so far.

s1monw commented:
I think the changes so far look great @imotov I left some minor comments

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksResponse.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```
readFrom and writeto can be removed as they only call super?

s1monw commented on core/src/main/java/org/elasticsearch/action/support/tasks/BaseTasksRequest.java:
```diff
@@ -95,6 +96,22 @@ public final Request nodesIds(String... nodesIds) {
+     *
+     * By default tasks with any ids are returned.
+     */
+    public long taskId() {
```
can this be getTaskId and setTaskId and maybe not return this please?

imotov replied:
This is ActionRequest and this is how we handle parameters in all our ActionRequests. I always hated this convention and I would love to use getTaskId and setTaskId there, but it's a convention. If we are officially going away from this convention, I will gladly change all task-management request to have standard setters and getters. Just let me know.

s1monw commented on core/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:
```diff
@@ -124,6 +126,22 @@ private NodeTasksResponse nodeOperation(NodeTaskRequest nodeTaskRequest) {
+        if (request.taskId() != BaseTasksRequest.ALL_TASKS) {
+            // we are only checking one task, we can optimize it
+            Task task = taskManager.getTask(request.taskId());
+            if (task != null && request.match(task)) {
```
when can this be null? can we assert the state that we are in here?

imotov replied:
The task id is specified by a user and by the time it reaches this node the task with the given id might be already gone or the user might have specified a wrong id to start with and this task might have never existed. In both cases taskManager would return null. I was thinking about throwing a TaskNotFoundException or something like this in this case, but at the end decided to just silently ignore the request for now. Not sure what would be a good user experience here.

s1monw replied:
I think an exception would be the right thing todo we should never hide stuff like this. Maybe you can add a generic `ResourceNotFoundException` that we can later use in other places as well to  replace stuff like `AliasMissingExcepiton` etc. that has a `404` response code?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * An action can override this method to be notified when the task is cancelled
+     */
+    protected void doCancel(String reason) {
```
this is unused do we need it? can we maybe make this class final?

imotov replied:
I am envisioning task to be used in two ways. An action can ask the task periodically, "am I cancelled yet?", or it can subscribe itself for cancellation notification and be more proactive in the way it handles it. The doCancel method support the second way of doing thing. This pattern is implemented in CancellableTaskWithNotification in one of the tests.

s1monw replied:
I have concerns here. I think the task should not carry such a responsibility. I think if we need something like this then the taskmanager should allow registration of event listeners that get notified on cancel?

s1monw commented on core/src/main/java/org/elasticsearch/tasks/Task.java:
```diff
@@ -104,4 +103,18 @@ public long getParentId() {
+    /**
+     * Called by the task manager to notify the task that it was registered
+     */
+    protected void onRegistered() {
```
these are also unused? can we remove them?

imotov replied:
Good point, we don't need these anymore!

s1monw commented:
left some minors LGTM otherwise

imotov commented:
@s1monw I push a new version. I have removed doCancel, but I cannot make it final because I need to make it extensible for other services to add task features. For example reindex API will need an ability to add throttling on the top of cancelling. I have also fixed a race condition in TransportCancelTasksAction, that I found while running additional tests.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksAction.java:
```diff
@@ -0,0 +1,46 @@
+import org.elasticsearch.client.ElasticsearchClient;
+
+/**
+ * Action for retrieving a list of currently running tasks
```
Copy and paste leftover.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -0,0 +1,73 @@
+    private String reason = DEFAULT_REASON;
+
+    /**
+     * Get information from nodes based on the nodes ids specified. If none are passed, information
```
Copy and paste leftover.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public boolean match(Task task) {
+        return super.match(task) && task instanceof CancellableTask;
```
If you ask to cancel a non-cancellable task will it silently skip it or throw an error back to the user?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequestBuilder.java:
```diff
@@ -0,0 +1,34 @@
+import org.elasticsearch.client.ElasticsearchClient;
+
+/**
+ * Builder for the request to retrieve the list of tasks running on the specified nodes
```
Copy and paste leftover.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -0,0 +1,273 @@
+/**
+ * Transport action that can be used to cancel currently running cancellable tasks.
+ * <p>
+ * For a task to be cancellable it has to return an instance of {@link CancellableTask} from {@link }
```
Missing link?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -0,0 +1,273 @@
+        return true;
+    }
+
+    private void setBanOnNodes(String reason, CancellableTask task, Set<String> nodes, BanLock banLock) {
```
s/set/send/?

nik9000 replied:
Er, probably not. But a bit confusing name because it looks like a typo.

s1monw commented:
LGTM
